### PR TITLE
Add Macedonian layout

### PR DESCRIPTION
**Pull request by boriseto**
In this PR I've gone by the advice of some other devs about how to go on. I've tried the layouts by copy-pasting them in my file system so I would make sure they are correct.
Even though I don't have anything to add in the autocomplete DB, it's still a step forward to have the needed letters and signs.

### CODE REVIEW

**Comment by boriseto** [comment]
Okay, not sure why, but the MK layout doesn't show when trying to install it in terminal. Any help appreciated.

**Comment by Fuseteam** [comment]
I've been thinking it may be myspell-mk perhaps remove that for the time being from debian/control

**Comment by boriseto** [comment]
Okay, ***maybe*** I needed to install the keyboard from terminal as well besides changing to the repo? One of the steps that isn't important at all. :D
screenshot20201118_163036678

**Comment by Fuseteam** [comment]
Yay

**Comment by Flohack74** [comment]
@boriseto is this ready for merge?

**Comment by boriseto** [comment]
> @boriseto is this ready for merge?

Yeap, it doesn't have autocomplete, but at least we have the layout and it's usable.

**Comment by Flohack74** [comment]
Ok then I add it to OTA-15 and lets see if we can merge it for that release.

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
Thanks for returning, this is looking good! Just one thing of note.

This keyboard has a very strange behavior on my devices... It pushes the Shift and Backspace buttons off screen.

image

I wonder if another layout with many keys, like the new Thai layout, has more instructions for how this should be handled?

**Comment by boriseto** [comment]
Ah, didn't test it on default scaling, had to move to Android for a while until I solved a few things with work.
Anyway, the Thai are using an extra row for the letters.
For this, I would actually like to change the size of the shift and backspace key to be with the size of the letters (it's the default setting that we have on almost every other OS for the Macedonian language) - with eleven buttons per row.
I couldn't find a way to do that, the only thing I could think of was adding leftside/rightside true to that row.

**Comment by Fuseteam** [comment]
@boriseto the keys for backspace and shift are defined in [the keys folder](https://github.com/ubports/keyboard-component/tree/xenial/qml/keys), you can perhaps make an copy of those keys and modify and use those instead of the standard ones in this layout

**Comment by Flohack74** [comment]
One last thing, can you please add your language in https://github.com/ubports/keyboard-component/blob/xenial/qml/keys/languages.js in the right alphabetical place? And can you build it locally, and also check in the resulting pot file in the po subfolder? Thanks!

**Comment by boriseto** [comment]
Sorry for the later reply, super busy at work currently and had to go back to Android for a while for the exchange email protocol, but I will add it to my tablet when I get to it again.

> @boriseto the keys for backspace and shift are defined in [the keys folder](https://github.com/ubports/keyboard-component/tree/xenial/qml/keys), you can perhaps make an copy of those keys and modify and use those instead of the standard ones in this layout

I'll take a look, thank you.

> One last thing, can you please add your language in https://github.com/ubports/keyboard-component/blob/xenial/qml/keys/languages.js in the right alphabetical place? And can you build it locally, and also check in the resulting pot file in the po subfolder? Thanks!

Ehm, about that, I thought that the Korean language is out of order, cause by code (MK), it comes after the letter L as I'm aware, succeeded by N? But I will and get back to you when I get my phone to UBtouch again.

**Comment by Flohack74** [comment]
I did not take a look I am just looking for someone who can help sorting and updating this file ^^

**Comment by boriseto** [comment]
Oh, okay, I can update it. Personally I wasn't sure why it was on that position, but yeah.
I just flashed UBtouch again, so I'll give it a go tomorrow, it's a bit too late now.

**Comment by boriseto** [comment]
> One last thing, can you please add your language in https://github.com/ubports/keyboard-component/blob/xenial/qml/keys/languages.js in the right alphabetical place? And can you build it locally, and also check in the resulting pot file in the po subfolder? Thanks!

Wait, do you want it in alphabetical order by country code or the way it's built on the view?
Cause in the languages view, the order is by the actual naming (for example Macedonian is Македонски), and being with a different type of letters, not Latin, those languages would always be at the bottom (including the other ones like Japanese, Korean and so on).

**Comment by boriseto** [comment]
> Thanks for returning, this is looking good! Just one thing of note.
> 
> This keyboard has a very strange behavior on my devices... It pushes the Shift and Backspace buttons off screen.
> 
> image
> 
> I wonder if another layout with many keys, like the new Thai layout, has more instructions for how this should be handled?

photo_2020-12-20_21-45-18

Well that was an easy fix after going through some files. Just defined the { width } as a property to take the size of the default keys (since it should always be that size at least for the Macedonian layout).

I'm waiting on the build and I'll reinstall it just to check if that part works.
I've tested it with default scaling as well, it looks perfect.

**Review by boriseto — COMMENTED** [review]
The changes have been made and commited. There's a screenshot as well. My vscode extension acted up for this review, sorry.

**Review by UniversalSuperBox — DISMISSED** [review]
Looks good to me. You @Flohack74?

File: plugins/mk/qml/Keyboard_mk.qml [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2013 Canonical Ltd.
```
Comment (mariogrip) [inline plugins/mk/qml/Keyboard_mk.qml:2]: We might want to change this

File: plugins/mk/qml/Keyboard_mk_email.qml [hunk 2]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2013 Canonical Ltd.
```
Comment (mariogrip) [inline plugins/mk/qml/Keyboard_mk_email.qml:2]: and this

File: plugins/mk/qml/Keyboard_mk_url.qml [hunk 3]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2013 Canonical Ltd.
```
Comment (mariogrip) [inline plugins/mk/qml/Keyboard_mk_url.qml:2]: ^

File: plugins/mk/qml/Keyboard_mk_url_search.qml [hunk 4]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2013 Canonical Ltd.
```
Comment (mariogrip) [inline plugins/mk/qml/Keyboard_mk_url_search.qml:2]: ^

**Review by mariogrip — DISMISSED** [review]
LGTM :+1: 

But we may want to change the copyright

File: plugins/mk/qml/Keyboard_mk_url_search.qml (hunk 4 above)
Comment (Fuseteam) [reply plugins/mk/qml/Keyboard_mk_url_search.qml:2]: to what tho?

**Comment by boriseto** [comment]
Are there actions required on my side for the functionality specifically?
I can help with the other stuff, but I guess it should be connected with another issue, not this one?

**Comment by UniversalSuperBox** [comment]
@boriseto, yes, you may want to update the copyright dates and names in the files you've created. They should be copyright 2021 either yourself or the UBports Foundation, if you wish to assign your copyright to the Foundation. Either is acceptable. Add these lines anew rather than replacing the old Canonical copyright. So at the end you'll end up with "copyright 2013 canonical" and on the next line "copyright 2021 ..."

File: plugins/mk/qml/Keyboard_mk_url_search.qml (hunk 4 above)
Comment (boriseto) [reply plugins/mk/qml/Keyboard_mk_url_search.qml:2]: Since I was a bit too busy and don't even have the machine nearby, @Fuseteam did the work for me. :)

**Review by UniversalSuperBox — APPROVED** [review]
I've installed this on my device and taken it for a test drive. While I don't know how to type this language, all of the given layouts seem to work properly without any unexpected layout changes between different types of input boxes.

Thank you very much. I'll merge this, then create the MR that will add it to the image.

**Comment by UniversalSuperBox** [comment]
I've merged https://gitlab.com/ubports/core/seeds/ubuntu-touch/-/merge_requests/25, which will add the macedonian keyboard to the image. It should be available in tomorrow's devel images and the next rc image.

**Pull request merged**